### PR TITLE
Upload artifacts in CI

### PR DESCRIPTION
**Pull request by andrewbranch**
This will let another process get the artifact URL and run arethetypeswrong on it. (Related: #66678)

### CODE REVIEW

File: scripts/ghostbuster.js [hunk 1]
```diff
@@ -65,7 +65,7 @@ function getAllHeaders() {
             const indexContent = readFileSync(index, "utf-8");
             let parsed;
             try {
-                parsed = hp.parseHeaderOrFail(indexContent);
+                parsed = hp.parseHeaderOrFail(index.pathname, indexContent);
```
Comment (andrewbranch) [inline scripts/ghostbuster.js:68]: Drive-by fix for https://github.com/microsoft/DefinitelyTyped-tools/pull/720

**Review by sandersn — APPROVED** [review]

**Comment by andrewbranch** [comment]
I would also appreciate a review from @jakebailey because every time I touch GitHub Actions I do something wrong and he figures it out before it even runs 😄

**Comment by jakebailey** [comment]
I'm a little confused by this (not the action stuff in general); what part of ATTW is this for? Checks in this repo? Shouldn't those be regular CI checks without needing artifacts? Or am I misinterpreting something? Is this just starter code and more will come later?

I think my only concern is the storage space of these artifacts; DT has a lot of CI runs which will all incur the cost of hosting these temporary packages, when it seems like they should only be needed within the scope of a single run, such that you could produce them on the fly in the same job as is checking them?

**Comment by andrewbranch** [comment]
This is related to what we were talking about a week or two ago—we can’t enable true checks in GHA right now because there’s no neutral conclusion. My original plan was to have individual tarballs uploaded and add support to the attw web app for reading an arbitrary tarball URL. Unfortunately, what we’re going to get out of this is a tarball of tarballs. I can still add support for that fairly easily, and/or have dt-mergebot unpack it and run the analysis and put the results in the header comment.

Re storage: artifacts have a default retention period of 90 days and that can be configured in repo settings.

**Comment by jakebailey** [comment]
Ah, right, tgz for ATTW linked to by the bot.

I guess I was imagining something different, e.g. running ATTW in CI and then uploading a static report instead, kinda like the package-size report on the TS repo or the preview site uploads (where the artifact is the "end result" so to speak). That way we can explicitly mark certain packages as skipping ATTW so that we could in fact make it a requirement someday.

**Comment by andrewbranch** [comment]
Hm, I guess I can do that.

> That way we can explicitly mark certain packages as skipping ATTW so that we could in fact make it a requirement someday.

Don’t understand how this is dependent on which asset gets uploaded, though

**Comment by jakebailey** [comment]
If you upload the package only, there's nothing in CI that can fail, right? It's purely advisory info that you have to go to the ATTW site to view and check? Or is there a CI component that will come in later?

I was basing my comment on the latter case being an eventuality, in that if we're going to check something in CI, then we would theoretically not need to have any artifacts at all anymore as it's all already in the job. Though if you wanted to view it on ATTW instead of in the CI, you'd need to upload the tgz to view on ATTW if that view is better.

**Comment by jakebailey** [comment]
Mainly I was just trying to get a view of the whole picture to make sure it would work out.

As it stands now, I think this will mean any change to say, `@types/node`, will cause us to build and upload every package which depends on node?

**Comment by andrewbranch** [comment]
Yeah, eventually I want to replace this step with one that just runs `attw` and can fail, but for now, something has to be uploaded so that another run on pull_request_target or workflow_dispatch or dt-mergebot can post the results in a comment, which requires running with secrets.

I think this will mean any change to say, `@types/node`, will cause us to build and upload every package which depends on node?

No, just Node itself. Dependents aren’t considered in the script. (Also, I forgot to do this, but I need to filter out ones like `@types/node` that aren’t for an npm implementation package.)

**Review by jakebailey — APPROVED** [review]
> Yeah, eventually I want to replace this step with one that just runs `attw` and can fail, but for now, something has to be uploaded so that another run on pull_request_target or workflow_dispatch or dt-mergebot can post the results in a comment, which requires running with secrets.

Makes sense, though I think I'd just be running ATTW within the untrusted context, and then upload its results for later posting instead, saving artifact room and making it less likely to cause a security hole. But if you want to link to these tarballs from the site for now, then this PR seems fine for now.

> No, just Node itself. Dependents aren’t considered in the script. (Also, I forgot to do this, but I need to filter out ones like `@types/node` that aren’t for an npm implementation package.)

I think I saw `getAffectedPackagesFromDiff` and got scared 😅 but you are just pulling `changedPackages` and not `dependentPackages`.

**Comment by andrewbranch** [comment]
Running the analysis and uploading the result is fine. It would be really convenient to upload the tarballs if I could get individual npm-ready tarballs as artifacts, but I don’t think it’s possible to do that when multiple could exist. https://github.com/actions/upload-artifact/issues/331

**Comment by jakebailey** [comment]
Yeah, the inability to upload raw artifacts is extremely annoying. Would love for TS CI to upload tarballs for every PR but they'd be double wrapped and so are useless to package managers :(

**Comment by andrewbranch** [comment]
Waiting on the monorepo effort to either conclude or stall, since if it goes through, this will be much easier.

**Comment by jakebailey** [comment]
Old but attw is added and doesn't need this anymore.
